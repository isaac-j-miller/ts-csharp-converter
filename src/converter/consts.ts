export const jsDocNumberTypes = [
  "decimal",
  "float",
  "double",
  "sbyte",
  "byte",
  "short",
  "ushort",
  "int",
  "uint",
  "long",
  "ulong",
  "nint",
  "nuint",
] as const;

export const primitiveTypeNames = [
  "string",
  "String",
  "number",
  "Number",
  "boolean",
  "Boolean",
  "object",
  "any",
  "undefined",
  "null",
  "unknown",
  "symbol",
  "Symbol",
  ...jsDocNumberTypes,
] as const;

export const CONSTS_KEYWORD = "__consts__" as const;
