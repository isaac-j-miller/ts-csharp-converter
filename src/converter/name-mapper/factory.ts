import { assertNever } from "src/common/util";
import {
  CamelOutputMapper,
  SnakeOutputMapper,
  PascalOutputMapper,
  KebabOutputMapper,
  normalize,
  parseNormalized,
} from "./mappers";
import { CasedString, CasingConvention, NameOutputMapper, NameMapperFunction } from "./types";

export function getOutputMapper<T extends CasingConvention>(target: T): NameOutputMapper<T> {
  switch (target) {
    case CasingConvention.SnakeCase:
      return SnakeOutputMapper as NameOutputMapper<T>;
    case CasingConvention.PascalCase:
      return PascalOutputMapper as NameOutputMapper<T>;
    case CasingConvention.CamelCase:
      return CamelOutputMapper as NameOutputMapper<T>;
    case CasingConvention.KebabCase:
      return KebabOutputMapper as NameOutputMapper<T>;
    default:
      assertNever(target);
  }
  throw new Error(`Could not find mapping for target: ${target}`);
}

export function getNameMapper<TTarget extends CasingConvention>(
  target: TTarget
): NameMapperFunction<TTarget> {
  const fromWords = getOutputMapper(target);
  return (str: string): CasedString<TTarget> => {
    const normalized = normalize(str);
    const parsed = parseNormalized(normalized);
    const newName = fromWords(parsed);
    return newName;
  };
}
