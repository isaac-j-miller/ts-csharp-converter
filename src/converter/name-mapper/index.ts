export * from "./types";
export * from "./mappers";
export * from "./factory";
