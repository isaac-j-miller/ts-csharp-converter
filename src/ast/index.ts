import {
  ForEachDescendantTraversalControl,
  InterfaceDeclaration,
  Node,
  Project,
  PropertySignature,
  SourceFile,
  SyntaxKind,
  Type,
  TypeAliasDeclaration,
} from "ts-morph";
import { TypeRegistry } from "src/converter/registry";
import {
  TypeRegistryType,
  TypeRegistryUnionType,
} from "src/converter/registry-types";
import { IRegistryType, PrimitiveTypeName } from "src/converter/types";
import { SyntheticSymbol } from "src/converter/synthetic/symbol";

export class AstTraverser {
  private project: Project;
  private entrySourceFile: SourceFile;
  private registry: TypeRegistry;
  constructor(entrypoint: string, tsconfigPath: string) {
    this.project = new Project({
      tsConfigFilePath: tsconfigPath,
    });
    this.entrySourceFile = this.project.getSourceFileOrThrow(entrypoint);
    this.registry = new TypeRegistry();
  }
  private processNode<T extends SyntaxKind>(kind: T, node: Node) {
    console.debug(`Got kind: ${SyntaxKind[kind]}`);
    // TODO: do something here
    return;
  }
  private processInterfaceDeclaration(node: InterfaceDeclaration) {}

  private createType(
    name: string,
    node: Node,
    asType: Type,
    internal: boolean = false
  ): IRegistryType {
    const symbolToUse =
      asType.getSymbol() ??
      asType.getAliasSymbol() ??
      new SyntheticSymbol(name, asType);
    if (asType.isUnion()) {
      const unionTypes = asType.getUnionTypes();
      if (unionTypes.every((unionType) => unionType.isStringLiteral())) {
        const members = unionTypes.map((member) =>
          member.getLiteralValueOrThrow()
        );
        const unionRegType = new TypeRegistryUnionType(
          name,
          symbolToUse,
          members as string[],
          internal
        );
        console.debug(`Adding type ${name} to registry`);
        this.registry.addType(unionRegType);
        return unionRegType;
      }
    }
    const regType = new TypeRegistryType(
      this.registry,
      name,
      symbolToUse,
      internal
    );
    const propertySignatures = asType.getApparentProperties();
    propertySignatures.forEach((property) => {
      const propertyName = property.getName();
      const isOptional = property.isOptional();
      const propertyType = property.getTypeAtLocation(node);
      const isArray = propertyType.isArray();
      const propertyText = propertyType.getText();
      const propertyTypeSymbol = propertyType.getSymbol();
      const arrayElemType = propertyType.getArrayElementType();
      if (name === "BlahBlah") {
        console.debug(name);
      }
      const getPrimitivePropertyType = (): PrimitiveTypeName | undefined => {
        if (propertyType.isString()) {
          return "string";
        }
        if (propertyType.isNumber()) {
          return "number";
        }
        if (propertyType.isBoolean()) {
          return "boolean";
        }
        if (propertyType.isAny()) {
          return "any";
        }
        return;
      };
      const primitiveType = getPrimitivePropertyType();
      if (propertyType.isTypeParameter()) {
        const genericParamName = propertyType.getSymbolOrThrow().getName();
        regType.addProperty(
          propertyName,
          {
            genericParamName,
            isGenericReference: true,
          },
          {
            isArray,
            isOptional,
          }
        );
      } else if (primitiveType) {
        const primitiveSymbol = this.registry
          .getType(primitiveType)!
          .getSymbol();
        regType.addProperty(propertyName, primitiveSymbol, {
          isArray,
          isOptional,
        });
      } else if (propertyTypeSymbol) {
        if (isArray) {
          const arrayElemTypeSymbol = arrayElemType!.getSymbol();
          if (arrayElemTypeSymbol) {
            regType.addProperty(propertyName, arrayElemTypeSymbol, {
              isArray,
              isOptional,
            });
          } else {
            console.debug(
              `Creating anonymous type ${name}${propertyName}Class`
            );
            const anon = this.createType(
              `${name}${propertyName}Class`,
              node,
              arrayElemType!,
              true
            );
            regType.addProperty(propertyName, anon.getSymbol(), {
              isArray,
              isOptional,
            });
          }
        } else {
          regType.addProperty(propertyName, propertyTypeSymbol, {
            isArray,
            isOptional,
          });
        }
      } else {
        const fromText = this.registry.findTypeBySymbolText(propertyText);
        if (fromText) {
          regType.addProperty(propertyName, fromText.getSymbol(), {
            isArray,
            isOptional,
          });
        } else {
          console.warn(
            `No symbol for property ${propertyName} on type ${name}!`
          );
          const primitiveSymbol = this.registry.getType("object")!.getSymbol();
          regType.addProperty(propertyName, primitiveSymbol, {
            isArray,
            isOptional,
          });
        }
      }
    });
    const genericParameters = asType.getTypeArguments();
    genericParameters.forEach((param) => {
      const v = (param.getSymbol() ?? param.getAliasSymbol())?.getName();
      if (!v) {
        console.debug(`Unable to find generic param name`);
        return;
      }
      regType.addGenericParameter(v);
    });
    console.debug(`Adding type ${name} to registry`);
    this.registry.addType(regType);
    return regType;
  }
  private processTypeAliasDeclaration(node: TypeAliasDeclaration) {
    const name = node.getName();
    const asType = node.getType();
    return this.createType(name, node, asType);
  }
  private traverseNode(node: Node) {
    node.forEachDescendant((node, traversal) => {
      const kind = node.getKind();
      // console.debug(`Traversing node with kind ${SyntaxKind[kind]}`);
      switch (kind) {
        // case SyntaxKind.ConditionalType:
        case SyntaxKind.TypeAliasDeclaration:
          this.processTypeAliasDeclaration(node.asKindOrThrow(kind));
          break;
        case SyntaxKind.InterfaceDeclaration:
          this.processInterfaceDeclaration(node.asKindOrThrow(kind));
          break;
        // case SyntaxKind.LiteralType:
        case SyntaxKind.MappedType:
        case SyntaxKind.EnumDeclaration:
          // case SyntaxKind.NumericLiteral:
          // case SyntaxKind.PropertySignature:
          // case SyntaxKind.TupleType:
          // case SyntaxKind.TypeLiteral:
          // case SyntaxKind.TypeParameter:
          // case SyntaxKind.TypeReference:
          this.processNode(kind, node);
          break;
        // case SyntaxKind.UnionType:
        //     break;
        case SyntaxKind.ImportDeclaration:
        case SyntaxKind.ExportDeclaration: {
          const sourceFile = node
            .asKindOrThrow(kind)
            .getModuleSpecifierSourceFileOrThrow();
          this.traverseNode(sourceFile);
          break;
        }
        default:
          break;
      }
    });
  }
  traverse() {
    this.traverseNode(this.entrySourceFile);
  }
  createNamespace(name: string) {
    return this.registry.toNamespace(name);
  }
}
