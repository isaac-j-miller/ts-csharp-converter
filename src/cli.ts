import "source-map-support/register"
import { ArgumentParser } from "argparse";
import { convertTypescriptToCSharp } from ".";
import {
  CasingConvention,
  NameMapperConfig,
  NameType,
} from "./converter/name-mapper";
import { toCasingConvention } from "./converter/name-mapper/util";

type RunConfigBase = {
  entrypoint: string;
  tsconfigPath: string;
  outputFile: string;
  namespace: string;
};

type RunConfig = RunConfigBase & {
  classNameSourceCasing?: string;
  classNameTargetCasing?: string;
  propertyNameSourceCasing?: string;
  propertyNameTargetCasing?: string;
  enumMemberSourceCasing?: string;
  enumMemberTargetCasing?: string;
};

type CompleteConfig = RunConfigBase & {
  nameMapperConfig: NameMapperConfig;
};

const argParser = new ArgumentParser();
argParser.add_argument("--entrypoint", {
  required: true,
  dest: "entrypoint",
});
argParser.add_argument("--tsconfig-path", {
  required: true,
  dest: "tsconfigPath",
});
argParser.add_argument("--output-file", {
  required: true,
  dest: "outputFile",
});
argParser.add_argument("--output-namespace", {
  required: true,
  dest: "namespace",
});
argParser.add_argument("--class-name-source-casing", {
  required: false,
  dest: "classNameSourceCasing",
});
argParser.add_argument("--class-name-target-casing", {
  required: false,
  dest: "classNameTargetCasing",
});
argParser.add_argument("--property-name-source-casing", {
  required: false,
  dest: "propertyNameSourceCasing",
});
argParser.add_argument("--property-name-target-casing", {
  required: false,
  dest: "propertyNameTargetCasing",
});
argParser.add_argument("--enum-member-source-casing", {
  required: false,
  dest: "enumNameSourceCasing",
});
argParser.add_argument("--enum-member-target-casing", {
  required: false,
  dest: "enumNameTargetCasing",
});
const runConfigToCompleteRunConfig = (c: RunConfig): CompleteConfig => {
  const {
    classNameSourceCasing,
    classNameTargetCasing,
    propertyNameSourceCasing,
    propertyNameTargetCasing,
    enumMemberSourceCasing: enumNameSourceCasing,
    enumMemberTargetCasing: enumNameTargetCasing,
    ...rest
  } = c;
  const nameMapperConfig: NameMapperConfig = {
    transforms: {
      [NameType.DeclarationName]: {
        input:
          toCasingConvention(classNameSourceCasing) ??
          CasingConvention.PascalCase,
        output:
          toCasingConvention(classNameTargetCasing) ??
          CasingConvention.PascalCase,
      },
      [NameType.PropertyName]: {
        input:
          toCasingConvention(propertyNameSourceCasing) ??
          CasingConvention.CamelCase,
        output:
          toCasingConvention(propertyNameTargetCasing) ??
          CasingConvention.PascalCase,
      },
      [NameType.EnumMember]: {
        input:
          toCasingConvention(enumNameSourceCasing) ??
          CasingConvention.CamelCase,
        output:
          toCasingConvention(enumNameTargetCasing) ??
          CasingConvention.PascalCase,
      },
    },
  };
  return {
    ...rest,
    nameMapperConfig,
  };
};
const getRunConfig = () => {
  const args: RunConfig = argParser.parse_args(process.argv);
  return runConfigToCompleteRunConfig(args);
};

async function main() {
  const { entrypoint, tsconfigPath, outputFile, namespace, nameMapperConfig } =
    getRunConfig();
  await convertTypescriptToCSharp(
    entrypoint,
    tsconfigPath,
    outputFile,
    namespace,
    nameMapperConfig
  );
}

void main();
