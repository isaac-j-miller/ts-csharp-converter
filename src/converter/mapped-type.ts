import { LoggerFactory } from "src/common/logging/factory";
import { Node, SyntaxKind, Type } from "ts-morph";
import { PrimitiveTypeName } from "./types";
import { getFinalSymbol } from "./util";

function isMappedType(type: Type): boolean {
  return type.getApparentProperties().length === 0;
}
export function getIndexAndValueType(
  node: Node
): [
  [Type | PrimitiveTypeName | undefined, Node | undefined],
  [Type | PrimitiveTypeName | undefined, Node | undefined]
] {
  const logger = LoggerFactory.getLogger("mapped-type-inferrer");
  const type = node.getType().getApparentType();
  if (type.getStringIndexType()) {
    return [
      ["string", undefined],
      [type.getStringIndexType(), undefined],
    ];
  }
  if (type.getNumberIndexType()) {
    return [
      ["number", undefined],
      [type.getNumberIndexType(), undefined],
    ];
  }
  if (!type.isObject()) {
    return [
      [undefined, undefined],
      [undefined, undefined],
    ];
  }
  if (!isMappedType(type)) {
    return [
      [undefined, undefined],
      [undefined, undefined],
    ];
  }
  const symbol = type.getAliasSymbol() ?? type.getSymbol();
  if (!symbol) {
    return [
      [undefined, undefined],
      [undefined, undefined],
    ];
  }
  const declaration = symbol.getDeclarations()[0];
  if (!declaration) {
    return [
      [undefined, undefined],
      [undefined, undefined],
    ];
  }
  let inBrackets = false;
  let afterColon = false;
  let afterSemicolon = false;
  const keyItems: Node[] = [];
  const valueItems: Node[] = [];
  const declarationDescendantToUse = declaration.getFirstDescendantByKind(SyntaxKind.MappedType);
  if (!declarationDescendantToUse) {
    return [
      [undefined, undefined],
      [undefined, undefined],
    ];
  }
  const descendants = declarationDescendantToUse.getChildren();
  for (const descendant of descendants) {
    const kind = descendant.getKind();
    switch (kind) {
      case SyntaxKind.OpenBracketToken:
        afterSemicolon = false;
        inBrackets = true;
        break;
      case SyntaxKind.CloseBracketToken:
        inBrackets = false;
        afterColon = false;
        break;
      case SyntaxKind.ColonToken:
        afterColon = true;
        break;
      case SyntaxKind.SemicolonToken:
        afterSemicolon = true;
        break;
      case SyntaxKind.TypeReference:
      case SyntaxKind.TypeAliasDeclaration:
      case SyntaxKind.TypeLiteral:
      case SyntaxKind.ArrayLiteralExpression:
      case SyntaxKind.StringKeyword:
      case SyntaxKind.StringLiteral:
      case SyntaxKind.NumberKeyword:
      case SyntaxKind.NumericLiteral:
      case SyntaxKind.BooleanKeyword:
      case SyntaxKind.LiteralType:
      case SyntaxKind.AnyKeyword:
      case SyntaxKind.ObjectKeyword:
      case SyntaxKind.ObjectLiteralExpression:
      case SyntaxKind.TypeParameter:
      case SyntaxKind.UnionType:
      case SyntaxKind.ArrayType:
      case SyntaxKind.TupleType:
      case SyntaxKind.InferType:
      case SyntaxKind.ConditionalType:
        if (afterSemicolon) {
          break;
        }
        if (inBrackets) {
          keyItems.push(descendant);
        } else if (afterColon) {
          valueItems.push(descendant);
        }
        break;
      default:
        break;
    }
  }

  const detectedIndex: [Type | undefined | PrimitiveTypeName, Node | undefined] = [
    undefined,
    undefined,
  ];
  const detectedValue: [Type | undefined | PrimitiveTypeName, Node | undefined] = [
    undefined,
    undefined,
  ];
  const nodeSymbol = node.getSymbol();
  const finalSynbol = nodeSymbol ? getFinalSymbol(nodeSymbol).getName() : "<anon>";
  if (keyItems.length === 1) {
    const toUse = keyItems[0];
    const asTypeParamDec = toUse.asKind(SyntaxKind.TypeParameter);
    if (asTypeParamDec) {
      detectedIndex[0] = asTypeParamDec.getConstraintOrThrow().getType();
      detectedIndex[1] = asTypeParamDec;
    } else {
      logger.warn("Key item not a type parameter declaration");
    }
  } else if (keyItems.length > 0) {
    logger.warn(`More than one key item detected for ${finalSynbol}:`, keyItems);
  }
  if (valueItems.length === 1) {
    const toUse = valueItems[0];
    detectedValue[0] = toUse.getType();
    detectedValue[1] = toUse;
  } else if (valueItems.length > 0) {
    logger.warn(`More than one value item detected for ${finalSynbol}:`, valueItems);
  }
  return [detectedIndex, detectedValue];
}
