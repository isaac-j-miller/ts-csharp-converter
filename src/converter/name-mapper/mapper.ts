import { getNameMapper } from "./factory";
import {
  NameMapperConfig,
  NameType,
  NameMapperFunction,
  CasedString,
  CasingConvention,
} from "./types";

type MapperMap = {
  [K in NameType]: NameMapperFunction;
};

export class NameMapper {
  private mappers: MapperMap;
  constructor(config: NameMapperConfig) {
    this.mappers = Object.keys(NameType)
      .filter((v) => !isNaN(Number(v)))
      .reduce((acc, curr) => {
        const asNameType = Number.parseInt(curr) as NameType;
        const transform = config.transforms[asNameType];
        const { input, output } = transform;
        acc[asNameType] = getNameMapper(input, output);
        return acc;
      }, {} as MapperMap);
  }
  transform(name: string, nameType: NameType): string {
    const mapperFn = this.mappers[nameType];
    const mapped = mapperFn(name as unknown as CasedString<CasingConvention>);
    return mapped.toString();
  }
}
