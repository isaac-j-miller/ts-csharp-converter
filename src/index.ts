import ejs from "ejs";
import { mkdir, writeFile, readFile } from "fs/promises";
import { join, resolve } from "path";
import { AstTraverser } from "./ast";
import { CasingConvention, NameMapperConfig } from "./converter/name-mapper";
import { NameMapper } from "./converter/name-mapper/mapper";

export { NameType } from "./converter/name-mapper/types";
export { CasingConvention, NameMapperConfig };

export async function convertTypescriptToCSharp(
  entrypoint: string,
  tsconfigPath: string,
  outputDir: string,
  namespaceName: string,
  config: NameMapperConfig,
  includeNodeModules: boolean,
  ignoreClasses: Set<string>
): Promise<void> {
  const mapper = new NameMapper(config);
  const traverser = new AstTraverser(entrypoint, tsconfigPath, includeNodeModules, ignoreClasses);
  traverser.traverse();
  const ns = traverser.createNamespace(namespaceName, mapper);
  const serializedNamespace = ns.serialize(mapper);
  const dir = join(resolve(outputDir), namespaceName);
  await mkdir(dir, {
    recursive: true,
  });
  const outputFile = join(dir, `${namespaceName}.cs`);
  const serializationUtilsPath = join(__dirname, "csharp/Source/SerializationUtils.cs");
  const csprojFilepath = join(__dirname, "csharp/Source/csproj.ejs");
  const serializationUtilsOutputPath = join(dir, "SerializationUtils.cs");
  const csprojOutputPath = join(dir, `${namespaceName}.csproj`);
  const serializationUtils = await readFile(serializationUtilsPath, { encoding: "utf-8" });
  await writeFile(outputFile, serializedNamespace, {
    encoding: "utf-8",
  });
  await writeFile(serializationUtilsOutputPath, serializationUtils, {
    encoding: "utf-8",
  });
  const csproj = await readFile(csprojFilepath, { encoding: "utf-8" });
  const templated = await ejs.render(csproj, { namespaceName }, { async: true });
  await writeFile(csprojOutputPath, templated, {
    encoding: "utf-8",
  });
}
