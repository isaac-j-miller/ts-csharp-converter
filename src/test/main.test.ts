import { AstTraverser } from "src/ast";
import { CasingConvention, NameMapper, NameType } from "src/converter/name-mapper";

describe("snapshot", () => {
  it("works", () => {
    const config = {
      transforms: {
        [NameType.DeclarationName]: {
          input: CasingConvention.PascalCase,
          output: CasingConvention.PascalCase,
        },
        [NameType.PropertyName]: {
          input: CasingConvention.CamelCase,
          output: CasingConvention.PascalCase,
        },
        [NameType.EnumMember]: {
          input: CasingConvention.CamelCase,
          output: CasingConvention.PascalCase,
        },
      },
    };
    const mapper = new NameMapper(config);
    const traverser = new AstTraverser("./src/test/index.ts", "./tsconfig.json", false, new Set());
    traverser.traverse();
    const ns = traverser.createNamespace("TestNamespace");
    const str = ns.serialize(mapper);
    expect(str).toMatchSnapshot();
  });
});
