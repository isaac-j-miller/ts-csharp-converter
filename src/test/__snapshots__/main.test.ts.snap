// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot works 1`] = `
"public namespace TestNamespace {
    using System;
    using System.Collections.Generic;
    public enum TestEnum1 {
        FooBar,
        Baz,
        Fooooo
    }
    public enum TestEnum2 {
        Value1 = 100,
        Value2,
        Value3
    }
    public class X {
        // whatever is a number property on X
        public Int Whatever { get; set; }
    }
    public class SomeInterface {
        public String? Foo { get; set; }
        public String Bar { get; set; }
    }
    public class SomeInterface2 {
        public String Foo2 { get; set; }
        public String Bar { get; set; }
        // whatever is a number property on X
        public Int Whatever { get; set; }
    }
    public class TypeWithEnums {
        public TestEnum1 Enum1 { get; set; }
        public TestEnum2 Enum2 { get; set; }
    }
    public static class GlobalVars {
        public const String FOO = \\"some string\\";
        public const String BAR = \\"some string as const\\";
        public const Int Whatever = 1;
        /**
        * an array of strings
        */
        public const String[] AStringArray = { \\"foo\\", \\"var\\", \\"aaaa\\" };
        public const Int[] ANumberArray = { 1, 2, 3, 4 };
        // a 2D int array
        public const Int[,] A2dArray = { { 1, 2, 3 }, { 2, 3, 4 }, { 4, 5, 6 } };
    }
    public class AAAAAAAA<T> {
        public T[] Vvv { get; set; }
    }
    public class Foo {
        // x is a number
        public Int X { get; set; }
        // comment comment comment
        public String Y { get; set; }
    }
    public enum Abc {
        A,
        B,
        C
    }
    public class GenericWithConstraints<T> where T: String {
        public T Str { get; set; }
    }
    public class GenericWithConstraints2<T> where T: Foo {
        public T Foo { get; set; }
    }
    public class GenericWithConstraints3<T> where T: Int {
        public T Digit { get; set; }
    }
    public class GenericWithConstraints4<T> where T: Object {
        public T Z { get; set; }
    }
    public class SomeGenericType<T> {
        public T Foo { get; set; }
        public Abc Bar { get; set; }
    }
    public class DefaultGeneric<Ta> {
        public Ta V { get; set; }
    }
    internal class SomeLargerTypexClass<U> {
        public U Foo { get; set; }
    }
    public class SomeLargerType<U> {
        public String A { get; set; }
        public Abc Y { get; set; }
        public Int V { get; set; }
        public SomeLargerTypexClass<U> X { get; set; }
    }
    public class Composite {
        public Int V { get; set; }
        public Abc Foo { get; set; }
        public Abc Bar { get; set; }
        // x is a number
        public Int X { get; set; }
        // comment comment comment
        public String Y { get; set; }
    }
    internal class BlahBlahxClass {
        // x is a number
        public Int X { get; set; }
        // comment comment comment
        public String Y { get; set; }
        public String Foo { get; set; }
        public Abc Bar { get; set; }
    }
    public class BlahBlah {
        public BlahBlahxClass[] X { get; set; }
        public SomeLargerType<DefaultGeneric<Int>>[] Y { get; set; }
    }
    public class ZZZ {
        public String Z { get; set; }
        public System.Collections.Generic.Dictionary<String, String> V { get; set; }
        public System.Collections.Generic.Dictionary<String, DefaultGeneric<Int>> Y { get; set; }
        public System.Collections.Generic.Dictionary<String, SomeGenericType<Abc>> X { get; set; }
    }
    public class Bs<T> : System.Collections.Generic.Dictionary<String, DefaultGeneric<T>> {

    }
    public enum SomeStupidSubEnum {
        FooBar,
        Baz
    }
    public enum SomeStupidSubEnum2 {
        Fooooo = 2,
        Value2 = 101
    }
    public class DumbType {
        public Double Num { get; set; }
        public Int Dumber { get; set; }
        public String Un { get; set; }
        public Object Absurd { get; set; }
    }
    internal class IndexTypeaClass {
        public Abc Foo { get; set; }
    }
    public class IndexType {
        public System.Collections.Generic.Dictionary<String, String> Abc { get; set; }
        public IndexTypeaClass A { get; set; }
    }
    public class GenericIndexType<T> where T: System.Collections.Generic.Dictionary<String, DefaultGeneric<Object>> {
        public IndexType I { get; set; }
        public T V { get; set; }
    }
    public class AnotherType {
        public String[,,] X { get; set; }
        public Int[,,,,] N { get; set; }
    }
    public class GenericArrayType {
        public SomeGenericType<Composite>[] J { get; set; }
        public AnotherType[,] K { get; set; }
    }
    public class GenericArrayType2<V> {
        public V[,] SomeProperty { get; set; }
        public DefaultGeneric<V>[] AnotherProperty { get; set; }
    }
    public class NumbersWithJsDoc {
        /**
        * @type {int}
        */
        public Int IntProperty { get; set; }
        /**
        * @type {float}
        */
        public Double FloatProperty { get; set; }
    }
    public class ConsumesInterface {
        public Int Integer { get; set; }
        public Int Float { get; set; }
    }
    public enum TestEnum3 {
        X,
        Y,
        Z
    }
    public class UnionIndex {
        public Int? A { get; set; }
        public Int? B { get; set; }
        public Int? C { get; set; }
    }
    public class UnionIndex2 {
        public Bool A { get; set; }
        public Bool B { get; set; }
        public Bool C { get; set; }
    }
    public class GenericUnionIndex<T, V> where T: Object : System.Collections.Generic.Dictionary<T, V> {

    }
    public class GenericUnionIndex2<T> where T: Object : System.Collections.Generic.Dictionary<T, String> {

    }
    public class GenericUnionIndex3<T> where T: Object : System.Collections.Generic.Dictionary<T, Object> {

    }
    public class GenericUnionIndex4<T> where T: Object : System.Collections.Generic.Dictionary<T, Object> {

    }
    public class GenericConsumer : System.Collections.Generic.Dictionary<String, Int> {

    }
    // this is a comment that should show up on the GenericConsumerConsumer class
    public class GenericConsumerConsumer {
        // g comment
        public GenericConsumer G { get; set; }
    }
    public class ATuple : System.Tuple<String, Int> {

    }
    public class BTuple : System.Tuple<Double, Double, Double> {

    }
    public class CTuple : System.Tuple<String, Double, UnionIndex2, Object> {

    }
    public class TupleMap : System.Collections.Generic.Dictionary<String, BTuple> {

    }
    public class GenericTupleMap<T, U, V> : System.Tuple<T, U, V> {

    }
    public class GenericTupleMap2<T, V> : System.Tuple<T, V[]> {

    }
    public class TupleMap2 : System.Tuple<String, String[], Int[,,]> {

    }
    public class TupleMap3 : System.Tuple<ATuple[], BTuple[,]> {

    }
    public class Tuple1 : System.Tuple<String, Double, SomeGenericType<Double>, DefaultGeneric<Double>, SomeGenericType<Abc>> {

    }
    public class A {
        public Int Foo { get; set; }
    }
    public class B {
        public String Bar { get; set; }
    }
    // Warning: This class might not have been generated correctly. Source to check /Users/imiller/code/personal/csharp-converter/src/test/subfolder/conditional.ts:10
    // Source code:
    // 
    // export type Inverse<T extends A | B> = T extends A ? B : A;
    public class Inverse<T> where T: Object {

    }
    public class ConsumesConditional {
        public B Prop1 { get; set; }
    }
    public class ConditionalMappedType<T, V> where T: String : System.Collections.Generic.Dictionary<String, Object> {

    }
    internal class ConsumesConditionalMappedTypexClass {
        public B A { get; set; }
        public B B { get; set; }
        public B C { get; set; }
    }
    internal class ConsumesConditionalMappedTypeyClass {
        public ConsumesConditional A { get; set; }
        public ConsumesConditional B { get; set; }
        public ConsumesConditional C { get; set; }
    }
    public class ConsumesConditionalMappedType {
        public ConsumesConditionalMappedTypexClass X { get; set; }
        public ConsumesConditionalMappedTypeyClass Y { get; set; }
    }
}"
`;
